design-proposals/amd-sev.md

```diff
@@ -0,0 +1,245 @@
+
+### Step #1 Validation of the platform
+
+The guest owner should get the assets listed bellow from the platform owner:
+
+- the `Platform Diffie-Hellman (PDH)` key exported from the SEV hardware
+- the complete key chain (up to the `root of trust`) used to sign the `PDH`
+- the OVMF image
+
+On a node where KubeVirt is running, the privileged `virt-handler` pod can export the required assets in advance to a `kubevirt-private` directory. Whenever a new VM requiring SEV is launched, the directory can be mounted to the `virt-launcher` pod in `read-only` mode. A user can then invoke the `virtctl` tool to download the certificate chain for validation (for that an additional command will be needed, e.g. `virtctl fetchcertchain ...`).
+
+In order to ensure that the platform is not compromised the guest owner needs to validate the `PDH` key using the provided chain. For that `sev-tool --validate_cert_chain ...` can be used (AMD root certificate should be downloaded from the web for verification). Additionally the UEFI image should be considered 'trustable' by the guest owner.
+
+### Step #2 Establishing secure channel with the platform
+
+To establish a secure communication channel with the platform a guest owner needs to provide a launch blob with the parameters of the session and the certificate derived from `PDH`. This is currently not in the scope of KubeVirt as it requires the guest owner to expose the secrets. It should be performed 'offline' by the guest owner (e.g. by using `sev-tool --generate_launch_blob ...`).
+
+The blob and certificate can be passed via `session` and `dhCert` parameters of the libvirt domain xml. This however brings a limitation: the guest owner needs to know in advance what node the VM will be scheduled on (which is usually not the case). Therefore additional handling is required in order to mitigate that in KubeVirt.
+
+Initially the guest owner can provide empty `dhCert` and `session` parameters in the VMI spec:
+
+```yaml
+spec:
+  domain:
+    launchSecurity:
+      sev:
+        dhCert: ""
+        session: ""
+```
+
+The implementation in `virt-launcher` in turn can 'postpone' the creation of the domain until the parameters are specified. Sidecar hooks functionality can be used to modify the domxml after a VMI has been scheduled. New `virtctl` commands (e.g. `virtctl launchblob ...`) can be introduced to support the process and to trigger propagation of the parameters.
+
+To run a SEV VM on a specific node the parameters can be specified directly in the VMI spec:
+
+```yaml
+spec:
+  nodeSelector:
+    kubernetes.io/hostname: "node01"
+  domain:
+    launchSecurity:
+      sev:
+        dhCert: <base64>
+        session: <base64>
+```
+
+To run a VM without going through the attestation process at all the parameters can be omited:
+
+```yaml
+spec:
+  domain:
+    launchSecurity:
+      sev: {}
+```
```

Comment:
I think there is value in having just simple implementation. Even if a user cannot do the attestation at the moment its still beneficial that the VM workload is run in confidential environment as it increases the level of data protection.

The PR https://github.com/kubevirt/kubevirt/pull/6166 enables SEV support in KubeVirt but does not touch the attestation topic. Maybe we could focus on that one first and then gradually proceed with enhacning the functionality. I more than welcome such an approach.

Reply:
I consent here with @vladikr  that we should at least have a clear plan to add the attestation process relatively fritcion-less later. @vasiliy-ul do I understand the flow correctly

 * VM gets scheduled
 * Users fetch the host cred chain, generate the blob and inject it
 * VM starts in pause mode and the measurement takes place
 * Users verify the measurment and allow by injecting the final secret the actual start
 
I guess it would be great if we could come up with an overall-flow of something like this:

```
kubectl create -f def.yaml
kubectl virt sev attestate #goes through all the steps as above and informs users of the progress
```

It would be great if we can provide for all relevant steps in the flow proper subresources (not necessarily one subresource for each step).

I would love to see clear apis here where interaction is needed and no things like entering pods to read some files from somewhere.

In addition we can add additionl debugging tooling subcommands like

```
kubectl virt sev showchain
```

and so forth.

Reply:
> I consent here with @vladikr that we should at least have a clear plan to add the attestation process relatively fritcion-less later.

No objections here. Though for me it seems that the two topics (launching and attesting) can be handled somewhat independently. Simply launching a SEV VM (without attestation) is purely a technical process (it needs just some configuration tweaks, validations and VMI spec API). Attestation may affect the launching in a sense that e.g. with attestation we need to start the VM paused. But that is IMHO an independent flow and should not affect the scenario without attestation.

> @vasiliy-ul do I understand the flow correctly

Yes, that is absolutely correct. And I basically agree with the rest about the subresources and APIs. I will probably look more into that direction further.

Reply:
Sounds good to me.

Reply:
Sounds great. We can start with the simple launch flow and then come back a refine the attestation.